tools/pubsys/src/aws/validate_ami/mod.rs

```diff
@@ -0,0 +1,799 @@
+    if let Some(write_results_path) = &validate_ami_args.write_results_path {
+        // Filter the results by given status, and if no statuses were given, get all results
+        info!("Writing results to file");
+        let filtered_results = validation_results.get_results_for_status(
+            validate_ami_args
+                .write_results_filter
+                .as_ref()
+                .unwrap_or(&vec![
+                    AmiValidationResultStatus::Correct,
+                    AmiValidationResultStatus::Incorrect,
+                    AmiValidationResultStatus::Missing,
+                ]),
+        );
+
+        // Write the results as JSON
+        serde_json::to_writer_pretty(
+            &File::create(write_results_path).context(error::WriteValidationResultsSnafu {
+                path: write_results_path,
+            })?,
+            &filtered_results,
+        )
+        .context(error::SerializeValidationResultsSnafu)?;
+    }
+
+    Ok(validation_results)
+}
+
+/// Validates EC2 images in a single region, based on a Vec (ImageDef) of expected images
+/// and a HashMap (AmiId, ImageDef) of actual retrieved images. Returns a HashSet of
+/// AmiValidationResult objects.
+pub(crate) fn validate_images_in_region(
+    expected_images: Vec<ImageDef>,
+    actual_images: &HashMap<AmiId, ImageDef>,
+    region: &Region,
+) -> HashSet<AmiValidationResult> {
+    let mut results = HashSet::new();
+
+    // Validate all expected images, creating an AmiValidationResult object
+    for mut image in expected_images {
+        // If the image is expected to be public, the specific launch permissions are irrelevant
+        // and were not retrieved for the actual image
+        if image.public {
+            image.launch_permissions = None;
+        }
+        results.insert(AmiValidationResult::new(
+            image.id.clone(),
+            image.clone(),
+            actual_images.get(&image.id).map(|v| v.to_owned()),
+            region.clone(),
+        ));
+    }
+
+    results
+}
+
+type RegionName = String;
+type AmiId = String;
+
+/// Parse the file holding image values. Return a HashMap of Region mapped to a vec of ImageDefs
+/// for that region.
+pub(crate) async fn parse_expected_amis(
+    expected_amis_path: &PathBuf,
+) -> Result<HashMap<Region, Vec<ImageDef>>> {
+    // Parse the JSON file as a HashMap of region_name, mapped to a JSON value
+    let expected_amis: HashMap<RegionName, serde_json::Value> = serde_json::from_reader(
+        &File::open(expected_amis_path.clone()).context(error::ReadExpectedImagesFileSnafu {
+            path: expected_amis_path,
+        })?,
+    )
+    .context(error::ParseExpectedImagesFileSnafu)?;
+
+    // If the value for each Region is a JSON object instead of an array, wrap it in an array
+    let vectored_images = expected_amis
+        .into_iter()
+        .map(|(region, value)| {
+            let image_values = match value.as_array() {
```

Comment:
Can we write a test for this behavior? As an alternative tip, you can get Serde to parse this for you "for free" via `#[serde(untagged)]` on an enum.

```rust
enum ImageDef {
    Image(Image),
    ImageList(Vec<Image>),
}

impl ImageDef {
    fn images(&self) -> Vec<Image> {
        match self {
            // etc
        }
    }
}
```